momento-functions/examples/turbopuffer-search.rs

```diff
@@ -0,0 +1,215 @@
+//! ```
+
+use std::time::Duration;
+
+use log::LevelFilter;
+
+use momento_functions::{WebError, WebResponse, WebResult};
+use momento_functions_host::{cache, encoding::Json, web_extensions::headers};
+use momento_functions_log::LogMode;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+
+#[derive(Deserialize, Debug)]
+struct Request {
+    query: String,
+    topk: Option<usize>,
+    include_attributes: Option<Vec<String>>,
+}
+
+#[derive(Deserialize, Debug)]
+struct EmbeddingResponse {
+    data: Vec<EmbeddingData>,
+}
+#[derive(Deserialize, Serialize, Debug)]
+struct EmbeddingData {
+    embedding: Vec<f32>,
+    index: usize,
+}
+
+// Default to 30 second caching time for queries in Momento
+const DEFAULT_TTL_SECONDS: u64 = 30;
+
+momento_functions::post!(search);
+fn search(Json(request): Json<Request>) -> WebResult<WebResponse> {
+    let headers = headers();
+    setup_logging(&headers)?;
+
+    let Request {
+        query,
+        topk,
+        include_attributes,
+    } = request;
+    // Transform our human-readable text into embeddings via OpenAI
+    let embeddings = get_cached_query_embedding(query)?;
+    // Default to top 5 results if not provided
+    let topk = topk.unwrap_or(5);
+    // Default to empty list if not provided
+    let include_attributes = include_attributes.unwrap_or_default();
+
+    // These are passed in as environment variables when creating the function
+    let turbopuffer_api_key = format!(
+        "Bearer {}",
+        std::env::var("TURBOPUFFER_API_KEY").unwrap_or_default()
+    );
+    let turbopuffer_endpoint = std::env::var("TURBOPUFFER_ENDPOINT").unwrap_or_default();
+
+    let result = momento_functions_host::http::post(
+        turbopuffer_endpoint,
+        [
+            ("Authorization".to_string(), turbopuffer_api_key.to_string()),
+            ("Content-Type".to_string(), "application/json".to_string()),
+            ("Accept".to_string(), "*/*".to_string()),
+            (
+                "User-Agent".to_string(),
+                "momento-turbobuffer-example".to_string(),
+            ),
+        ],
+        json!({
+            "rank_by": ["vector", "ANN", embeddings],
+            "top_k": topk,
+            "include_attributes": include_attributes,
+        }),
+    );
+    log::debug!("Turbopuffer response: {result:?}");
+    match result {
+        Ok(response) => {
+            if response.status != 200 {
+                let message = format!(
+                    "Failed to search documents: {}",
+                    String::from_utf8(response.body).unwrap_or_default(),
+                );
+                return Ok(WebResponse::new()
+                    .with_status(response.status)
+                    .with_body(json!({
+                        "message": message,
+                    }))?);
+            }
+            Ok(WebResponse::new()
+                .with_status(200)
+                .with_body(Json(response))?)
```

Comment:
passing the turbopuffer response straight back could maybe make sense, but wrapping it in a Json encoder seems like it will obfuscate the response? Maybe you want to `.with_headers(content_type!("application/json")).with_body(response)`? I think they send you json back anyway, yeah?

Reply:
Oh yeah that makes more sense, let's do that

Reply:
Actually yeah, I think the `rows` is really all we care about:

```
[{"dist":0.45686936,"id":"13956"},{"dist":0.46907777,"id":"10876"},{"dist":0.46907777,"id":"25487"},{"dist":0.47382885,"id":"34738"},{"dist":0.49211055,"id":"33617"}]
```

Reply:
https://github.com/momentohq/functions/pull/2/commits/4c6b9a78ebffda38a2d39b0ddae350628da7d342